ui/tests/integration/allocation-row-test.js

```diff
@@ -120,3 +121,67 @@ test('Allocation row shows warning when it requires drivers that are unhealthy o
       assert.ok(find('[data-test-icon="unhealthy-driver"]'), 'Unhealthy driver icon is shown');
     });
 });
+
+test('when an allocation is not running, the utilization graphs are omitted', function(assert) {
+  this.setProperties({
+    context: 'job',
+    enablePolling: false,
+  });
+
+  // All non-running statuses need to be tested
+  ['pending', 'complete', 'failed', 'lost'].forEach(clientStatus =>
+    this.server.create('allocation', { clientStatus })
+  );
+
+  this.store.findAll('allocation');
+
+  return wait().then(() => {
+    const allocations = this.store.peekAll('allocation');
+    return waitForEach(
+      allocations.map(allocation => () => {
+        this.set('allocation', allocation);
+        this.render(hbs`
+            {{allocation-row
+              allocation=allocation
+              context=context
+              enablePolling=enablePolling}}
+          `);
+        return wait().then(() => {
+          const status = allocation.get('clientStatus');
+          assert.notOk(find('[data-test-cpu] .inline-chart'), `No CPU chart for ${status}`);
+          assert.notOk(find('[data-test-mem] .inline-chart'), `No Mem chart for ${status}`);
+        });
+      })
+    );
+  });
+});
+
+// A way to loop over asynchronous code. Can be replaced by async/await in the future.
+const waitForEach = fns => {
+  let i = 0;
+  let done = () => {};
+
+  // This function is asynchronous and needs to return a promise
+  const pending = new Promise(resolve => {
+    done = resolve;
+  });
+
+  const step = () => {
+    // The waitForEach promise and this recursive loop are done once
+    // all functions have been called.
+    if (i >= fns.length) {
+      done();
+      return;
+    }
+    // Call the current function
+    var promise = fns[i]() || resolve(true);
+    // Increment the function position
+    i++;
+    // Wait for async behaviors to settle and repeat
+    promise.then(() => wait()).then(step);
+  };
+
+  step();
+
+  return pending;
+};
```

Comment:
This is basically a crappy implementation of async/await. It also will be deleted once I migrate everything to async/await, which will be part of the 3.0 upgrade epic. 

Addressing comments in no particular order...

1. Yes, it needs to be called in order, so `Promise.all` won't work
2. I want to block on each promise, so `reduce` won't work (which is also why it's using the awkward recursive pattern, in the style of `requestAnimationFrame`)
3. Good catch on the different Promise implementations. It's probably safe, but consistency is good so I switched to `rsvp.Promise`.
4. I am making use of the return values? The functions passed to `waitForEach` are promises (`wait`) that are used to block before calling `step` again. The promise returned from `waitForEach` is returned from the `test` function to make sure the test doesn't end early.
5. `wait` is necessary because of the render call.

Generally speaking, I'd rather not dig into this function too much. It's not my favorite code and as mentioned, it's shim test code that will be replaced by async/await.

Reply:
Ace sounds good to me!

> Good catch on the different Promise implementations. It's probably safe, but consistency is good so I switched to rsvp.Promise

After reviewing this this morning I found one in my own code just after 😂 . We should be able to tweak ESLint to warn us of this shouldn't we?

> I am making use of the return values? The functions passed to waitForEach are promises (wait) that are used to block before calling step again. The promise returned from waitForEach is returned from the test function to make sure the test doesn't end early.

I think I meant that last promise, which returns a `true` I think? Doesn't matter, if this is going soon then it's all good.

Reply:
Hey, sorry misread that again last night just before I finished up for the day, it doesn't return `true`, it returns nothing. Just to clarify it was this bit:

https://github.com/hashicorp/nomad/blob/58b92d5b3c49da43aba5b9efa9a30ebd4aecf035/ui/tests/integration/allocation-row-test.js#L173

...reading the code `done` is `resolve` so it resolves `undefined` not `true`, and that is what I meant. Anyway no biggie, it's just I realized I misread that bit of code last night.

Also, took me a while to find, but I found this:

https://github.com/hashicorp/consul/pull/4454#discussion_r208423775

..which is the thing I referred to above about using `reduce` to do the same thing:

https://github.com/hashicorp/consul/blob/4e4eb053523066d13afeccc83b45fbe7ac956528/ui-v2/tests/lib/measure/getMeasure.js#L10-L24

This could be simplifed even more for your usecase, it's essentially:

```javascript
callbacks.reduce(function(prev, cb) {
  return prev.then(function() {
    return cb();
  });
}, Promise.resolve()).then(function() {
  console.log("I'm done!")
});
```
...you could probably shorten it more with arrow functions if you wanted to also, which makes for a nice one liner but is more difficult for the casual reader to read/figure out. You know personally I like to write `function`, maybe the below is easier for others?

```javascript
callbacks.reduce((prev, cb) => prev.then(() => cb()), Promise.resolve).then(console.log)

```

I know you are getting rid of this at some point, but thats ^ a good way to recur through promise returning callbacks with just ES5. Not entirely sure if it could replace what you have above, but I think it does.

Reply:
Gah! Just realized there's that native `Promise` in that `getMeasure`! 😂 

Need to look at getting ESLint to warn me of that!